### PR TITLE
Fix Params JSON Parsing

### PR DESCRIPTION
**Pull request by eliasjpr**
#### Depends on https://github.com/amberframework/granite-orm/pull/78 
☝️ (This needs to be merged in order for this branch to pass)

The framework should be able to parse JSON request body appropriately by returning a `JSON::Any` object.

Currently, there is a bug when trying to parse JSON parameters from the body of a request where nested JSON payload is not being parsed correctly

```crystal
# Given a JSON payload 
{ "test": "test", "address": { "city": "New York" }
# Should be accessed 
params["address"].as(Hash)["city"]
```
Params is now a custom Hash of the following signature `Hash(String | Symbol, Amber::Router::ParamsType)` or `Amber::Router::ParamsHash.new`

Issue https://github.com/amberframework/amber/issues/310

## UPDATED:

To give more context all parameters where being parsed using the **HTTP::Params.parse** method which returns an object of type **Hash(String, Array(String))** this caused issues when parsing JSON params from the body, It turns out that nested JSON keys where being parsed as a string. Basically, all `params` values are currently returning as a string. 

With the changes presented in this PR users can now define model objects with JSON mapping that should be evaluated correctly based on the mapping defined. So for instance:

```crystal
params["user_id"].as(Int64) => 123
params["address"].as(Hash).each ...
params["user"].as(User) => User.name 
# or 
User.from_json(params["user"]) => User.name
```
See https://github.com/amberframework/amber/pull/315/files#diff-707c9269bb28ed65da046758b2df5c48R156 
✋ @watzon

### CODE REVIEW

**Comment by watzon** [comment]
Just to clarify the bug affected/affects basically anything that's not a string; booleans, integers, floats, and nested objects.

**Comment by elorest** [comment]
What needed fixed with param and form parameters?

**Comment by eliasjpr** [comment]
Updated PR description

File: src/amber/router/params.cr [hunk 1]
```diff
@@ -1,129 +1,4 @@
-require "json"
-
 module Amber::Router
-  # The Params module will parse parameters from a URL, a form post or a JSON
```
Comment (elorest) [inline src/amber/router/params.cr]: Why was everything but 4 lines deleted from this file and copied to params_parser?

File: src/amber/router/params.cr [hunk 2]
```diff
@@ -1,129 +1,4 @@
...
-               else
-                 ""
-               end
-
-      HTTP::Params.parse(values) do |key, value|
-        params.add(key, value)
-      end
-    end
-  end
+  alias ParamsType = Nil | String | Int64 | Float64 | Bool | Hash(String, JSON::Type) | Array(JSON::Type)
```
Comment (elorest) [inline src/amber/router/params.cr]: I like this.

File: src/amber/router/params_parser.cr [hunk 3]
```diff
@@ -0,0 +1,127 @@
+require "json"
```
Comment (elorest) [inline src/amber/router/params_parser.cr]: I don't like creating a whole new file with mostly the same contents of params.cr. This screws up git history.

**Comment by elorest** [comment]
It's still named, "Fixes Parsing URL, Form, JSON Parameters" but I'm unclear what has been changed with parsing URL and Form parameters.

**Comment by elorest** [comment]
Now that params is a Hash instead of HTTP::Params we can include files in it as well which is nice.

File: src/amber/router/params.cr (hunk 1 above)
Comment (eliasjpr) [reply src/amber/router/params.cr]: The rename of the file is necessary here since it adds context on what the file does.

File: src/amber/router/params_parser.cr (hunk 3 above)
Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: I renamed the file to something more meaningful.

**Comment by elorest** [comment]
The changes that affect the generators and add validation code into controllers and models should be a separate PR.

**Comment by eliasjpr** [comment]
## Update Friday Oct 20 2017

Recently added another commit which addresses the issue with the generator_spec.cr:7 (Very useful spec).

The new changes modifies the app generator specifically the controller and models. 

**Controller**
The controller now generates the app using `params.validation` similar to Rails strong params. It adds a method to the Controller and overloads the Granite initializer and set_attributes methods. 

```crystal
# AnimalsController
private def animal_params
  params.validation do
    required("name") { |v| v.str? & !v.empty? }
  end
end
# animal_params.validate! returns a hash of `Hash(String | Symbol, Amber::Router::ParamsType)`
# only containing the attributes defined in the `params.validation`
animal = Animal.new(animal_params.validate!) 
```

**Model**
Models generate for the Granite::ORM now includes the following line `include Amber::Models::Granite` this will allow models to be populated with the controller params using the `params.validations`. The reason this is added here and not to the Granite::ORM repo is to keep Granite loosly couple from any framework, since the `params.validation` is specific to amber.

The module defines 2 methods an initializer and set attributes. See https://github.com/amberframework/amber/pull/315/files#diff-d748cfb908fdf84d3356f37cb13c2d0d 
 
```crystal
class Animal < Granite::ORM
  include Amber::Models::Granite
end
```

**Comment by eliasjpr** [comment]
This is now passing here https://travis-ci.org/amberframework/amber/builds/290448518

File: src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr [hunk 4]
```diff
@@ -1,6 +1,7 @@
 require "granite_orm/adapter/<%= @database %>"
 
 class <%= @name.capitalize %> < Granite::ORM::Base
+  include Amber::Models::Granite
```
Comment (drujensen) [inline src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr]: yuck, don't like this at all.  Anyway of avoiding special Granite logic in Amber?

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 5]
```diff
@@ -63,4 +63,14 @@ class <%= @name.capitalize %>Controller < ApplicationController
     end
     redirect_to "/<%= @name %>s"
   end
+
+  def <%= @name %>_params
+	params.validation do
+	  <% @fields.each do |field| %>
+	  <% unless field.cr_type() != "time" %>
+	  required(:<%= field.name %>) { |p| !p.as(<%= field.cr_type.capitalize %>) }
```
Comment (drujensen) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: I'm not following this logic.  Why would all the fields be required except times?

File: src/amber/models/granite.cr [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	  def initialize
+	  end
+
+	  def initialize(args : Hash(String | Symbol, Amber::Router::ParamsType))
+        set_attributes(args)
+  	  end
+
+	  def set_attributes(args : Hash(Symbol | String, Amber::Router::ParamsType))
+        args.each do |k, v|
+          cast_to_field(k, v.as(DB::Any))
```
Comment (drujensen) [inline src/amber/models/granite.cr]: instead of doing this here, can we cast it in the controller before passing into Granite?  Also, how does Crecto handle this?

File: src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr (hunk 4 above)
Comment (eliasjpr) [reply src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr]: We can have the initializer and set_attributes defined in Granite. I thought of that, but this is specific to Amber.

This will clean everything

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 5 above)
Comment (eliasjpr) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: This are the incoming URL/Form/Body params  that gets parsed. The time types are in reference to 'created_at' and 'updated_at' maybe it should be checking for those specific fields. 🤔

File: src/amber/models/granite.cr (hunk 6 above)
Comment (eliasjpr) [reply src/amber/models/granite.cr]: Well, the controller params does not care about the Model. The model should be responsible for casting to what it needs. 

Also the `cast_to_field` also live in Granite::ORM. We can easily just move the initialize and set_attributes to Granite

File: src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr (hunk 4 above)
Comment (drujensen) [reply src/amber/cli/templates/model/granite/src/models/{{name}}.cr.ecr]: ok,  I'm torn.  This feels wrong to me but I see your argument for putting it here.  Are you ok with moving it to granite?

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 5 above)
Comment (drujensen) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: yeah, I think you should iterate the `@visible_fields` since it takes care of which fields would be passed in the params.  There may be other fields that are `time` fields that are not created_at or updated_at.

File: src/amber/models/granite.cr (hunk 6 above)
Comment (drujensen) [reply src/amber/models/granite.cr]: I would prefer not requiring users to remember they need the included this module.  Can we move this to Granite?

**Comment by eliasjpr** [comment]
@drujensen this was my initial thought. I have remove the granite  file from Amber and opened this PR in Granite-ORM

Depends on https://github.com/amberframework/granite-orm/pull/78 (This needs to be merged first in order for this branch to pass)

This actually worked quite well and is clean with no dependencies between each other.

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 5 above)
Comment (eliasjpr) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Unfortunately `@visible_fields` is an array of only names (String) and I need the type as well.

File: src/amber/router/params_parser.cr [hunk 7]
```diff
@@ -1,28 +1,29 @@
 require "json"
 
 module Amber::Router
-  # The Params module will parse parameters from a URL, a form post or a JSON
+  alias ParamsType = Nil | String | Int64 | Float64 | Bool | Hash(String, JSON::Type) | Array(JSON::Type)
```
Comment (drujensen) [inline src/amber/router/params_parser.cr]: `Float32 | Int32 | Time | Bytes`?

File: src/amber/cli/templates/auth.cr [hunk 8]
```diff
@@ -25,7 +25,7 @@ module Amber::CLI
       @timestamp = Time.now.to_s("%Y%m%d%H%M%S")
       @primary_key = primary_key
       @visible_fields = @fields.reject(&.hidden).map(&.name)
```
Comment (drujensen) [inline src/amber/cli/templates/auth.cr]: we should remove this if its no longer being used.

Comment (eliasjpr) [reply src/amber/cli/templates/auth.cr]: **@visible_fields** is still being use in other places

File: spec/amber/router/context_spec.cr [hunk 9]
```diff
@@ -147,23 +147,24 @@ describe HTTP::Server::Context do
...
     headers = HTTP::Headers.new
     headers["Content-Type"] = "application/json"
-    body = "[\"test\",\"test2\"]"
+    body = %(["test", "test2"])
     request = HTTP::Request.new("POST", "/", headers, body)
 
     context = create_context(request)
 
-    context.params["_json"].should eq "[\"test\", \"test2\"]"
+    context.params["_json"].should eq ["test", "test2"]
```
Comment (veelenga) [inline spec/amber/router/context_spec.cr]: Might be handy: `%w(test test2)`

File: spec/amber/validations/params_spec.cr [hunk 10]
```diff
@@ -187,17 +187,25 @@ module Amber::Validators
...
     end
   end
 end
+
+def params_builder(body = "")
+  params = {} of String | Symbol => Amber::Router::ParamsType
+  HTTP::Params.parse(body).each do |key, value|
+    params[key] = value.as(Amber::Router::ParamsType)
+  end
+  params
```
Comment (veelenga) [inline spec/amber/validations/params_spec.cr]: [Enumerable.each_with_object](https://crystal-lang.org/api/0.23.1/Enumerable.html#each_with_object%28obj%2C%26block%29-instance-method) returns that object, so you don't need that extra line to return it.

```crystal
params = {} of String | Symbol => Amber::Router::ParamsType

HTTP::Params.parse(body).each_with_object(params) do |key, value|
  params[key] = value.as(Amber::Router::ParamsType)
end
```

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 11]
```diff
@@ -63,4 +63,14 @@ class <%= @name.capitalize %>Controller < ApplicationController
     end
     redirect_to "/<%= @name %>s"
   end
+
+  def <%= @name %>_params
+	params.validation do
+	  <% @fields.each do |field| %>
+	  <% unless ["created_at", "updated_at"].includes? field.name %>
```
Comment (veelenga) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: `%w(created_at updated_at).includes? ...`

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 12]
```diff
@@ -63,4 +63,14 @@ class <%= @name.capitalize %>Controller < ApplicationController
...
+	  required(:<%= field.name %>) { |p| !p.as(<%= field.cr_type.capitalize %>) }
```
Comment (veelenga) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Shouldn't it be `as?` instead of `as` ? `as` rises an exception which would result in 500 error code instead of validation error.

Comment (eliasjpr) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: I personally prefer the exception knowing that is not being cast to the type I am expecting, but it can be changed to `as?`, I guess  it depends in the use case. IMHO.

File: spec/amber/validations/params_spec.cr (hunk 10 above)
Comment (eliasjpr) [reply spec/amber/validations/params_spec.cr]: I tried this and got an error laving it as is for now

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 12 above)
Comment (veelenga) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: @eliasjpr imagine a view with a form that requires integer value. User enters "blah" and submits. In my understanding controller will throw exception and user will be responded with status code 500, instead of a flash message "you have to enter integer value". Am I wrong?

File: spec/amber/validations/params_spec.cr (hunk 10 above)
Comment (veelenga) [reply spec/amber/validations/params_spec.cr]: Sorry, this is the correct one:

```crystal
params = {} of String | Symbol => Amber::Router::ParamsType

HTTP::Params.parse(body).each_with_object(params) do |(key, value), params|
  params[key] = value.as(Amber::Router::ParamsType)
end

```

**Comment by eliasjpr** [comment]
@veelenga should be good now. Thanks for the nice suggestions!

**Comment by veelenga** [comment]
@eliasjpr i still have doubts regrading `as` method. I compiled amber on this branch and generated new app:

```sh

$ amber new testapp
$ cd testapp
$ crystal deps
$ amber generate scaffold Post title:string no:integer
$ amber w 
```

First of all there is a strange formatting in PostController:

```crystal
  def post_params
	params.validation do
	  
	  
	  required(:title) { |p| !p.as(String) }
	  
	  
	  
	  required(:no) { |p| !p.as(Int32) }
	  
	  
	  
	  
	  
	  
	end
  end
```

Second of all is it does not compile:

```crystal
in src/controllers/post_controller.cr:22: instantiating 'post_params()'

  post = Post.new(post_params.validate!)
                  ^~~~~~~~~~~

in src/controllers/post_controller.cr:75: can't cast String to Int32

   required(:no) { |p| !p.as(Int32) }

```

I think it still requires some improvements/fixes.

**Comment by eliasjpr** [comment]
Yes you’re right the post_params should not be, just params. I’ll clean/fix it up. Thanks

**Comment by eliasjpr** [comment]
@veelenga I completely had a misconception of `as()` I thought this was a way of casting

**Comment by veelenga** [comment]
@eliasjpr it is a way of casting, but with some type inference features:

```crystal
"string".as(Int32) #=> can't cast String to Int32 (compile error)
```

```crystal
if false
  a = 1
else
  a = "hello"
end

a.as(String) #=> ok 
a.as(Int32)  #=> cast from String to Int32 failed (runtime error, TypeCaseError)

```

**Comment by eliasjpr** [comment]
@veelenga yeah that looks more like unboxing. 

Anyways I made changes to this PR. The only workaround I found for the moment was to parse the JSON inside the validation predicate in params.validation. I also had to open a PR in Granite::ORM to add more support for JSON types. https://github.com/amberframework/granite-orm/pull/82

Ideally for the params we should create something similar to JSON.mapping so we can map values pre compilation and properly cast the values for each params. The benefits of this is that it will have strict type params hence stronger validations and speed since it does not need to go through the JSON.parse.

To have this branch pass it is necessary to merge this PR https://github.com/amberframework/granite-orm/pull/82

File: spec/spec_helper.cr [hunk 13]
```diff
@@ -1,12 +1,13 @@
...
-TEST_PATH    = "spec/support/sample"
-PUBLIC_PATH  = TEST_PATH + "/public"
-VIEWS_PATH   = TEST_PATH + "/views"
-TESTING_APP  = "./tmp/test_app"
-APP_TPL_PATH = "./src/amber/cli/templates/app"
-CURRENT_DIR  = Dir.current
+TEST_PATH     = "spec/support/sample"
+PUBLIC_PATH   = TEST_PATH + "/public"
+VIEWS_PATH    = TEST_PATH + "/views"
+TEST_APP_NAME = "test_app"
```
Comment (elorest) [inline spec/spec_helper.cr]: Why was this changed?

File: src/amber/cli/templates/auth.cr (hunk 8 above)
Comment (elorest) [reply src/amber/cli/templates/auth.cr]: @visible_fields is being used everywhere forms or controllers are being generated.

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 14]
```diff
@@ -19,7 +19,7 @@ class <%= @name.capitalize %>Controller < ApplicationController
   end
 
   def create
-    <%= @name %> = <%= @name.capitalize %>.new(params.to_h.select(<%= @visible_fields %>))
+  <%= @name %> = <%= @name.capitalize %>.new(<%= @name.downcase%>_params.validate!)
```
Comment (elorest) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Does `validate!` take into affect visible fields?

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 15]
```diff
@@ -63,4 +63,26 @@ class <%= @name.capitalize %>Controller < ApplicationController
     end
     redirect_to "/<%= @name %>s"
   end
+
+  def <%= @name %>_params
+    params.validation do
+      <%- @fields.each do |field| -%>
+      <%- unless %(created_at updated_at).includes? field.name -%>
```
Comment (elorest) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: This conditional doesn't seem to have a body to execute if true.

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 16]
```diff
@@ -63,4 +63,26 @@ class <%= @name.capitalize %>Controller < ApplicationController
...
+      <%- if %w(Int32 Int64).includes? field.cr_type -%>
+      required(:<%= field.name %>) { |p| !JSON.parse(p.as(String)).as_i.nil? }
+      <%- elsif %w(Bool).includes? field.cr_type -%>
+      required(:<%= field.name %>) { |p| JSON.parse(p.as(String)).as_bool }
```
Comment (elorest) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Why is everything being parsed as JSON? Most params won't be JSON.

**Comment by eliasjpr** [comment]
@elorest I would love to pair on this. I got burned out over the weekend on this. 

All params are now a Hash of ParamsType but most of them will be strings when parsed from the request, This doesn't allow to cast to the correct type. For instance `params["name"].as(Int32)` where name is a String, this throws an error because at the moment it was populated it was set as string. Using JSON.parse allows to cast to the correct type. 

> It would be ideal to be able to map the params to their correct type so when the request gets parsed there is no need for casting.

I understand that this does not look pretty `JSON.parse(p.as(String)).as_f.nil?` but is the only way to cast between types to the correct right now in Crystal.  

The PR looks big, not intended. I hope there is a better shorter way to do type casting in crystal but it seems that is not supported by the language at the moment.

**Comment by watzon** [comment]
Looks like I really opened a can of worms didn't I? Haha. That's one of the things that I feel like Crystal really needs, a [martialing system[(https://golang.org/pkg/encoding/json/) like that of golang.

File: src/amber/router/params_parser.cr [hunk 17]
```diff
@@ -1,28 +1,29 @@
 require "json"
 
 module Amber::Router
-  # The Params module will parse parameters from a URL, a form post or a JSON
+  alias ParamsType = Float32 | Int32 | Time | Bytes | Nil | String | Int64 | Float64 | Bool | Hash(String, JSON::Type) | Array(JSON::Type)
```
Comment (elorest) [inline src/amber/router/params_parser.cr]: alias ParamsType = (String | JSON::Type)

File: src/amber/router/params_parser.cr [hunk 18]
```diff
@@ -90,10 +91,10 @@ module Amber::Router
           case json = JSON.parse(body).raw
           when Hash
             json.each do |key, value|
-              params[key.as(String)] = value.to_s
+              params[key] = value.as(Amber::Router::ParamsType)
```
Comment (elorest) [inline src/amber/router/params_parser.cr]:
```cr
    def parse_json
      if body = request.body.not_nil!.gets_to_end
        if body.size > 2
          case json = JSON.parse(body).raw
          when Hash
            json.each do |key, value|
              case value.raw.class
              when String
                params[key.as(String)] = value.to_s
              else
                params[key.as(String)] = value
              end
            end
          when Array
            params["_json"] = json.to_s
          end
        end
      end
    end
```

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 16 above)
Comment (eliasjpr) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: I will be removing this `{ |p| !JSON.parse(p.as(String)).as_i.nil? }` to this `{ |p| !p.nil? }`

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 15 above)
Comment (eliasjpr) [reply src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: The intention is not to render the `created_at` and `updated_at` for params validations

File: spec/spec_helper.cr (hunk 13 above)
Comment (eliasjpr) [reply spec/spec_helper.cr]: We talked offline. 

This is to avoid the `split('/').last` on the test. See https://github.com/amberframework/amber/pull/315/files#diff-3269a1cd7605965238570f1b2aafc0a7L16

File: src/amber/router/params_parser.cr [hunk 19]
```diff
@@ -90,10 +91,10 @@ module Amber::Router
           case json = JSON.parse(body).raw
           when Hash
             json.each do |key, value|
-              params[key.as(String)] = value.to_s
+              params[key.as(String)] = value
```
Comment (elorest) [inline src/amber/router/params_parser.cr]: If JSONType is a string I think we should set it as a string otherwise just set it to value. WDYT?

Comment (elorest) [reply src/amber/router/params_parser.cr]:
```cr
          when Hash
            json.each do |key, value|
              case value.raw.class
              when String
                params[key.as(String)] = value.to_s
              else
                params[key.as(String)] = value
              end
            end
          when Array
            params["_json"] = json.to_s
          end
```

Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: JSON::Parse returns always a  JSON::Any

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr (hunk 14 above)
Comment (drujensen) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: Seems like we would want this indented 2 spaces

File: src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr [hunk 20]
```diff
@@ -63,4 +63,14 @@ class <%= @name.capitalize %>Controller < ApplicationController
     end
     redirect_to "/<%= @name %>s"
   end
+
+  def <%= @name %>_params
+    params.validation do
+      <%- @fields.each do |field| -%>
+      <%- unless %(created_at updated_at).includes? field.name -%>
+      required(:<%= field.name %>) { |p| p.to_s.empty? }
```
Comment (drujensen) [inline src/amber/cli/templates/scaffold/controller/granite/src/controllers/{{name}}_controller.cr.ecr]: This is failing for checkboxes that may not send a value in the params.

```bash
amber g scaffold post title:string body:text published:bool
```

When trying to create a new blog post, I get an error:
```
ERROR: Validation failed. [Amber::Validators::Error(@param="title", @value="This is a test", @message="Field title is required"), Amber::Validators::Error(@param="body", @value="Testing", @message="Field body is required"), Amber::Validators::Error(@param="published", @value="0", @message="Field published is required")] (Amber::Exceptions::Validator::ValidationFailed)
```

File: src/amber/router/params_parser.cr (hunk 19 above)
Comment (elorest) [reply src/amber/router/params_parser.cr]: Yeah. That's why I'm saying that if it's a string instead of a hash we should just use the string value as the value instead of json::any. If it's anything except the string then just set it to the Json any value. See above.

Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: screen shot 2017-10-25 at 11 11 13 am

**Comment by eliasjpr** [comment]
@drujensen and @elorest I have tested this generating a new test app with auth and model. This were the steps taken. This updates the Auth templates as well

1. make (Compile amber cli from Params branch to ensure is using the correct code)
2. amber n test_app
3. cd test_app
4. amber g scaffold Post title:string body:text published:bool
5. amber g auth user
6. Edit database.yml 
7. amber m up
8. shards install
9. amber w
10. Visit http://0.0.0.0:3000
11. Create account
12. Sign In
13. Create Post
14. Edit Post

File: src/amber/router/params_parser.cr [hunk 21]
```diff
@@ -106,7 +107,7 @@ module Amber::Router
         if filename.is_a?(String) && !filename.empty?
           files[upload.name] = Files::File.new(upload: upload)
```
Comment (eliasjpr) [inline src/amber/router/params_parser.cr]: For a separate PR this should be move to the params scope beware that the filename can collide with a parameter found in the body or url. Also `ParamsType` would need to include the union `Files::File`

Comment (elorest) [reply src/amber/router/params_parser.cr]: `upload.name` is just an a file input name. So sure it could collide but so could any other input field.

File: spec/amber/amber_spec.cr [hunk 22]
```diff
@@ -55,8 +54,6 @@ describe Amber do
       end
 
       it "retains environment.yml settings that haven't been overwritten" do # NOTE: Any changes to settings here remain for all specs run afterwards.
-        # This is a problem.
```
Comment (elorest) [inline spec/amber/amber_spec.cr]: Until settings is change to an instances instead of singleton this is still an "a problem".

File: src/amber/router/params_parser.cr (hunk 19 above)
Comment (elorest) [reply src/amber/router/params_parser.cr]: How is value an Array(JSON::Type) for you? The array should still be JSON::Any. In my tests it is.

File: src/amber/router/params_parser.cr [hunk 23]
```diff
@@ -1,28 +1,29 @@
 require "json"
 
 module Amber::Router
-  # The Params module will parse parameters from a URL, a form post or a JSON
+  alias ParamsType = String | JSON::Type
```
Comment (elorest) [inline src/amber/router/params_parser.cr]: Should this be JSON::Any?

File: src/amber/router/params_parser.cr (hunk 19 above)
Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: We should pair on that one, to give you more context. But in a nutshell JSON.parse will return a JSON::Type and since we are not ‘unboxing’ to a specific type it reads as the first type `maybe` (I can be completely wrong here). 

This might work `value.as(String).raw.is_a? String` now this part of the params parser only gets trigger if the request is Application/JSON, this works fine as it is. So I think we can leave it

File: src/amber/router/params_parser.cr (hunk 23 above)
Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: See https://crystal-lang.org/api/0.23.1/JSON/Type.html

Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: it will be spending  a lot of time instantiating a JSON::Any for each json param, instead of doing this ahead of time without knowing what params will be use in the request life-cycle the user can determine when he wants to parse out the JSON to the correct type or simply wrap it in a JSON::Any object. JSON::Any is a convenience wrapper to traverse the JSON.

Comment (elorest) [reply src/amber/router/params_parser.cr]: But JSON.parse returns JSON::Any not JSON::Type right? So isn't it already doing that work?

File: src/amber/router/params_parser.cr (hunk 19 above)
Comment (elorest) [reply src/amber/router/params_parser.cr]: `.raw` only exists on unions and abstracts though. So if it's an Array(JSON::Any) .raw won't exist. In my tests values from JSON.parse are always JSON::Any though. Let me know when you have time to pair on this. Hopefully we can get it merged today.

File: src/amber/router/params_parser.cr (hunk 23 above)
Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: JSON.parse return JSON::Type See https://github.com/crystal-lang/crystal/blob/e2a1389e8165fb097c785524337d7bb7b550a086/src/json.cr#L83

Comment (eliasjpr) [reply src/amber/router/params_parser.cr]: Disregard the my comment above since I was looking at the wrong method 😕

File: src/amber/router/params_parser.cr (hunk 19 above)
Comment (elorest) [reply src/amber/router/params_parser.cr]: Ok so I didn't know you could do parse_raw. I was basing this off of the old implementation. So this is what I should have said.

```cr
          when Hash
            json.each do |key, value|
             if value.is_a? String
                params[key.as(String)] = value.to_s
             else
                params[key.as(String)] = value
             end
            end
          when Array
            params["_json"] = json
          end
```

File: src/amber/router/params_parser.cr (hunk 23 above)
Comment (elorest) [reply src/amber/router/params_parser.cr]: Got it thanks.

**Comment by elorest** [comment]
@eliasjpr Why is this closed. I thought we were going to merge it?

**Comment by watzon** [comment]
@elorest I was wondering the same thing
